### PR TITLE
Move all failures into connectFailed and callFailed.

### PR DESCRIPTION
**Pull request by swankjesse**
I think applications might be simple by tracking successful events
separately from failure events.

### CODE REVIEW

File: okhttp/src/main/java/okhttp3/EventListener.java [hunk 1]
```diff
@@ -65,7 +65,7 @@ public EventListener create(Call call) {
    *
    * <p>This will be invoked only once for a single {@link Call}. Retries of different routes
    * or redirects will be handled within the boundaries of a single callStart and
-   * {@link #callEnd(Call, IOException)} pair.
+   * {@link #callEnd(Call)} pair.
```
Comment (yschimke) [inline okhttp/src/main/java/okhttp3/EventListener.java]: nit: may end with a callFailed

File: okhttp/src/main/java/okhttp3/EventListener.java [hunk 2]
```diff
@@ -86,15 +86,8 @@ public void dnsStart(Call call, String domainName) {
    * Invoked immediately after a DNS lookup.
    *
    * <p>This method is always invoked after {@link #dnsStart(Call, String)}.
```
Comment (yschimke) [inline okhttp/src/main/java/okhttp3/EventListener.java]: nit: not true

File: okhttp/src/main/java/okhttp3/EventListener.java [hunk 3]
```diff
@@ -128,33 +121,31 @@ public void secureConnectStart(Call call) {
    * Invoked immediately after a TLS connection was attempted.
    *
    * <p>This method is always invoked after {@link #secureConnectStart(Call)}.
```
Comment (yschimke) [inline okhttp/src/main/java/okhttp3/EventListener.java]: nit: not true

**Comment by yschimke** [comment]
The cleanup of try catch blocks in CallServerInterceptor is convincing.

File: okhttp-tests/src/test/java/okhttp3/RecordingEventListener.java [hunk 4]
```diff
@@ -230,10 +236,27 @@ public String getName() {
...
-    ConnectEnd(Call call, InetSocketAddress inetSocketAddress, Proxy proxy, Protocol protocol,
+    ConnectEnd(Call call, InetSocketAddress inetSocketAddress, Proxy proxy, Protocol protocol) {
+      super(call, inetSocketAddress, proxy, protocol);
+      this.inetSocketAddress = inetSocketAddress;
+      this.proxy = proxy;
+      this.protocol = protocol;
+    }
+
+    @Nullable @Override public CallEvent closes() {
+      return new ConnectStart(call, inetSocketAddress, proxy);
```
Comment (PromanSEW) [inline okhttp-tests/src/test/java/okhttp3/RecordingEventListener.java]: Costructor never returns `null` value, so `@Nullable` annotation is wrong

File: okhttp-tests/src/test/java/okhttp3/EventListenerTest.java [hunk 5]
```diff
@@ -114,6 +117,22 @@
     assertEquals(expectedEvents, listener.recordedEventTypes());
   }
 
+  @Test public void canceledCallEventSequence() throws IOException {
```
Comment (dave-r12) [inline okhttp-tests/src/test/java/okhttp3/EventListenerTest.java]: 👍

File: okhttp-tests/src/test/java/okhttp3/RecordingEventListener.java [hunk 6]
```diff
@@ -49,7 +49,7 @@ public void forbidLock(Object lock) {
       event = eventSequence.poll();
     }
     if (event == null) throw new AssertionError();
-    return (T) event;
+    return eventClass.cast(event);
```
Comment (dave-r12) [inline okhttp-tests/src/test/java/okhttp3/RecordingEventListener.java]: Nice

File: okhttp/src/main/java/okhttp3/EventListener.java (hunk 1 above)
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/EventListener.java]: Fixed.

File: okhttp/src/main/java/okhttp3/EventListener.java (hunk 2 above)
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/EventListener.java]: Fixed, `This method is invoked after {@link #dnsStart}`.

File: okhttp/src/main/java/okhttp3/EventListener.java (hunk 3 above)
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/EventListener.java]: Fixed!

File: okhttp-tests/src/test/java/okhttp3/RecordingEventListener.java (hunk 4 above)
Comment (swankjesse) [reply okhttp-tests/src/test/java/okhttp3/RecordingEventListener.java]: Fixed!

**Pull request merged**